SwatchTooltip component

supernova-at opened the pull request:
<!--
Thank you for your contribution!

Before submitting this pull request, please make sure you have read our Contribution Guidelines and your PR meets our contribution standards:
https://github.com/magento-research/pwa-studio/blob/master/.github/CONTRIBUTION.md

Please fill out as much information as you can about your PR to help speed up the review process.
If your PR addresses an existing GitHub Issue, please refer to it in the title or Additional Information section to make the connection.

We may ask you for changes in your PR in order to meet the standards set in our Contribution Guidelines. PRs that do not comply with our guidelines may be closed at the maintainers' discretion.

Feel free to remove this section before creating this PR.
-->

## Description
<!--- Provide a general summary of your changes in the Title above -->
<!--- Describe your changes in detail here: -->

This PR adds a new `SwatchTooltip` component and uses it in the `ProductOptions` views (product page and edit mini cart item).

It updates the closed #712.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here with the specific wording: "Closes #<issue>" -->
<!--- Using the above wording causes Github to automatically close the issue on merge. -->
Closes #700 .

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
There was a UAT ticket for providing these tooltips.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
`yarn test` and manually on the product pages (PDPs) and in the mini-cart edit item dialog.

## Screenshots (if appropriate):

![image](https://user-images.githubusercontent.com/13182778/53369953-048c2a00-3912-11e9-8058-93f54801f8f1.png)

## Proposed Labels for Change Type/Package
<!--- What types of changes does your code introduce? Let us know if this is a -->
<!--- BUG, FEATURE, DOCUMENTATION, or TEST change. -->
FEATURE

<!--- What packages are modified by this code? Let us know if this applies to -->
<!--- peregrine, pwa-buildpack, upward-js, upward-spec, venia-concept or pwa-devdocs -->
`venia-concept`

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the **CONTRIBUTING** document.
- [x] I have linked an issue to this PR.
- [x] I have indicated the change type and relevant package(s).
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All CI checks are green (linting, build/deploy, etc).
- [ ] At least one core contributor has approved this PR.

soumya-ashok commented:
Had a look at the implementation. The left and right padding around the label needs to be reduced, and the vertical padding increased. It also looks odd that the tip is off-center.

<img width="353" alt="screen shot 2019-02-28 at 10 16 50 am" src="https://user-images.githubusercontent.com/25960021/53580818-09372500-3b42-11e9-8efe-c081a926c5b1.png">

supernova-at commented on packages/venia-concept/src/components/ProductOptions/__tests__/__snapshots__/option.spec.js.snap:
```diff
@@ -14,30 +14,50 @@ exports[`renders Option component correctly 1`] = `
+      </span>
+      <button
+        className="root"
+        itemIndex={0}
```
> This unrecognized DOM attribute angers the React.

😄

supernova-at commented:
> Had a look at the implementation. The left and right padding around the label needs to be reduced, and the vertical padding increased. On second thought, off-center tip makes the tooltip look unbalanced.

Totally agree.

> Would centering the tip cause any issues with display when the tooltip needs to appear on the swatch at the left edge of the screen?

It would, yeah. Would you mind if we merged this now and I create a follow up ticket to address these concerns?

soumya-ashok commented:
@supernova-at Sure, as long as we fix the issues I'm fine with changes being addressed in a follow-up ticket.

supernova-at commented:
> As long as we fix the issues I'm fine with changes being addressed in a follow-up ticket.

@soumya-ashok Awesome - created #1024 for follow up.

supernova-at commented:
> we have an opportunity for an a11y win here

Any chance I could convince you to allow this in the followup ticket? 😄

sirugh commented:
>Any chance I could convince you to allow this in the followup ticket?

I don't think it's *that* much work to add. `role="tooltip"` and `aria-describedby="<id>"` are easy enough. The focus things are also easy I think, probably just removing z-index. Heck I've got downtime so I could look into it tomorrow if you want. Just let me know :D

jimbo commented:
> Had a look at the implementation. The left and right padding around the label needs to be reduced, and the vertical padding increased. On second thought, off-center tip makes the tooltip look unbalanced.

Agreed on all points.

supernova-at commented:
In progress....

- [x] The tooltip never receives focus. Focus stays on the textbox.
- [ ] The tooltip widget can shown via keyboard focus or by the onmouse event.
- [ ] The tooltip widget can be hidden by removing focus from the text box or by moving the mouse off the textbox.
- [ ] The tooltip widget can be hidden by pressing the Escape key
- [ ] The tooltip is only hidden via JavaScript and CSS selectors. If JavaScript is not available the tooltip is shown.
- [x] The element that serves as the tooltip container has role tooltip.
- [ ] The element that triggers the tooltip references the tooltip element with aria-describedby.

supernova-at commented:
PR Updated:
  * Incorporated accessibility changes
  * Updated #1024 , as I took care of some easy wins here as long as I was touching the code again anyway. All that's left in 1024 is the centering of the triangle, which warrants larger design discussions.

supernova-at commented on packages/venia-concept/src/components/SwatchTooltip/swatchTooltip.js:
```diff
@@ -0,0 +1,111 @@
+            // Update its accessibility accordingly.
+            const ariaEnhancedButton = React.cloneElement(child, {
+                'aria-describedby': this.uniqueId,
+                className: `${child.props.className} ${classes.trigger}`,
```
I'm a little hesitant about composing the classnames together like this, but I'm not sure why. Basically I just want to add this component's class onto whatever is already present.

For example, `root tile` becomes `root tile trigger`.

sirugh replied:
I'm a little confused about what the desired outcome is here. In practice this will check for the `classes.trigger` prop from the css we have created, except that in `swatchTooltip.css` there is no `trigger` class so this is `undefined`. Do we even need this class addition?

supernova-at replied:
Nice catch, thanks!

If I remember correctly it was adding the `margin` around the swatch, which I thought definitely didn't belong in the `SwatchTooltip`, so I refactored the CSS and the `trigger` class doesn't exist anymore.

I'll remove this.

supernova-at commented on packages/venia-concept/src/components/clickable.css:
```diff
@@ -4,7 +4,6 @@
     display: inline-flex;
     justify-content: center;
     line-height: 1;
-    padding: 0;
```
This is not really a necessary change, I can back it out if desired.

jimbo replied:
Clickable is essentially a mixin you would use for some shared styles between links and buttons of all kinds. You use it by writing `composes: clickable` in your component CSS. As such, it doesn't *change* the padding of a button to `0`, it *resets* the padding to `0`, like a default. Your component would override that padding as needed.

sirugh commented on packages/venia-concept/src/components/SwatchTooltip/swatchTooltip.js:
```diff
@@ -0,0 +1,111 @@
+    };
+
+    onKeyDown = event => {
+        if (event.key === 'Escape') {
```
```suggestion
        if (event.key === 'Escape' || event.key === 'Tab') {
```

This would also hide it if you are moused in and then press tab. This could present an issue wherein there are focusable children within the button, imagine some text within the swatch, and pressing `Tab` should move focus to the text and keep the tooltip above the swatch.

supernova-at replied:
This suggestion would help solve the problem of multiple tooltips showing at the same time. This occurs when one Swatch has focus and another is being moused over.

But it may not be worth the effort.

1. We'd have to add logic to `SwatchTooltip` to workaround the new issue that it would present, which would be duplicating a lot of what `blur` is supposed to tell us.
1. Since this isn't a form, users usually aren't navigating with keyboard and mouse.

We could add some sort of `TooltipController` component that ensures that only one tooltip is showing at a time, but that seems like overkill too.

jimbo replied:
Generally speaking, if the expected behavior is that only one tooltip can be open at a time, then it's not local state: it's shared state. So it sounds like this situation does call for a "controller". Context is good for this purpose.

sirugh replied:
At my last company we had a `FocusManager` component that handled this sort of thing. Probably a bit overkill for now.

supernova-at replied:
@jimbo great points all around.

> if the expected behavior is that only one tooltip can be open at a time

This is the piece I'm iffy about. I mean, it sounds right but it is also a pretty rare occurrence so I lean towards "out of scope" here. When this thing makes it into Peregrine we can definitely generalize and make improvements to it.

What do you think?

jimbo replied:
> When this thing makes it into Peregrine we can definitely generalize and make improvements to it.

Yeah, that's probably the right approach.

supernova-at commented:
PR Updated:
  * Removes erroneous `className` composition during `cloneElement`

jimbo commented on packages/venia-concept/src/components/SwatchTooltip/swatchTooltip.js:
```diff
@@ -0,0 +1,110 @@
+                onMouseLeave: over([
+                    child.props.onMouseLeave,
+                    this.onMouseLeave
+                ])
```
I've never seen this pattern before. Ordinarily, you would call the props callback from the class handler. Any reason we shouldn't do that here?

Also, by convention, class handlers are named `handleEvent` rather than `onEvent`.

```jsx
class Foo extends Component {
  handleClick = event => {
    const { id, onClick } = this.props

    // this component's business
    event.preventDefault()

    // its parent's business
    // note: event object is treated as private
    // callback gets only what it needs
    if (typeof onClick === "function") {
      onClick(id)
    }
  }

  render() {
    const { handleClick } = this

    return <button onClick={handleClick} />
  }
}
```

supernova-at replied:
Yeah that's a good call - the reason this particular case is different is because this `SwatchTooltip` component is the parent, not the child.

In other words:

```jsx
<SwatchTooltip>
  <Swatch />
</SwatchTooltip>
```

So for a given event handler - let's say `onClick` - `SwatchTooltip` (the parent) wants its handler executed while also not trampling on `Swatch` (the child)'s handler.

I suppose I could do something like

```js
    onMouseOver = () => {
        // parent business.
        this.setState({ isShowing: true });

        // child business.
        const child = this.getChild(); // some logic that uses React.Children to get the right one
        if (child.props.onMouseOver) {
          child.props.onMouseOver();
        }
    };
```

and repeat that pattern for each event handler.

I'm mostly just trying to avoid tightly coupling these components.

jimbo replied:
> I'm mostly just trying to avoid tightly coupling these components.

Why's that? It seems to me that the tooltip should be a child of the `Swatch`, rendered by `Swatch`, and pretty coupled to it.

supernova-at replied:
> Why's that? It seems to me that the tooltip should be a child of the Swatch, rendered by Swatch, and pretty coupled to it.

Good point, that would definitely make things much simpler here. In the future we may want a generic `Tooltip` component like this, but that's out of scope here.

jimbo commented on packages/venia-concept/src/components/SwatchTooltip/swatchTooltip.css:
```diff
@@ -0,0 +1,42 @@
+}
+
+/* The tooltip triangle that points down. */
+.tooltip:before {
```
Love me some pseudo-element triangles.

sirugh commented:
We're going to merge this with the caveat that we try to figure out how to invert the parent-child relationship of the ToolTip component.
